cli/deployment/config.go

```diff
@@ -0,0 +1,122 @@
+package deployment
+
+import (
+	"fmt"
+	"gopkg.in/yaml.v2"
+	"log"
+	"regexp"
+	"strings"
+)
+
+var /* const */ pattern = regexp.MustCompile(`\$\(out\.(?P<token>[-.a-zA-Z0-9]+)\)`)
+
+type Config struct {
+	Name        string
+	Project     string
+	Description string
+	Imports     []struct {
+		Name string
+		Path string
+	}
+	Resources []interface{}
+	file      string
+	data      string
+}
+
+func NewConfig(data string, file string) Config {
+	config := Config{
+		file: file,
+		data: data,
+	}
+
+	err := yaml.Unmarshal([]byte(data), &config)
+	if err != nil {
+		log.Fatalf("error: %v", err)
+	}
+	return config
+}
+
+func (c Config) findAllDependencies(configs map[string]Config) []Config {
+	refs := c.findAllOutRefs()
+	if refs != nil {
+		dependencies := map[string]Config{}
+		for _, ref := range refs {
+			fullName, _, _ := parseOutRef(ref)
+			dependency, found := configs[fullName]
+			if !found {
+				log.Fatalf("Could not find config for deployment = %s", fullName)
+			}
+			dependencies[fullName] = dependency
+		}
+		var result []Config
+		for _, dependency := range dependencies {
+			fmt.Printf("Adding dependency %s -> %s\n", c.FullName(), dependency.FullName())
+			result = append(result, dependency)
+		}
+		return result
+	}
+	return nil
+}
+
+func (c Config) Yaml() ([]byte, error) {
+	imports, typeMap := c.importsAbsolutePath()
+
+	tmp := struct {
+		Imports   interface{}
+		Resources interface{}
+	}{
+		Imports:   imports,
+		Resources: c.resources(typeMap),
+	}
+	return yaml.Marshal(tmp)
+}
+
+func (c Config) importsAbsolutePath() (imports interface{}, typeMap map[string]string) {
+	typeMap = map[string]string{}
+	for i, imp := range c.Imports {
+		newPath := AbsolutePath(c.file, imp.Path)
+		if newPath != c.Imports[i].Path {
+			typeMap[c.Imports[i].Path] = newPath
+		}
+		c.Imports[i].Path = newPath
+	}
+	return c.Imports, typeMap
+}
+
+func (c Config) resources(typeMap map[string]string) []interface{} {
+	if len(typeMap) > 0 {
+		for i := range c.Resources {
+			res := c.Resources[i].(map[interface{}]interface{})
+			if typeMap[res["type"].(string)] != "" {
+				res["type"] = typeMap[res["type"].(string)]
+			}
+		}
+	}
+	return c.Resources
+}
+
+func (c Config) FullName() string {
+	return c.Project + "." + c.Name
+}
+
+func (c Config) String() string {
+	return c.FullName()
+}
+
+func (c Config) findAllOutRefs() []string {
+	matches := pattern.FindAllStringSubmatch(c.data, -1)
+	result := make([]string, len(matches))
+	for i, match := range matches {
+		result[i] = match[1]
+	}
+	if len(result) == 0 {
+		return nil
```

Comment:
is it rule true not just for arrays but for struct objects as well? I mean if we'll return struct instance anyway with lets fields initialized with default values, isn't it error prone?

Reply:
To be clear, this function returns a slice rather than an array type.

This code review suggestion is incorrect on my part, probably due to some mental gymnastics I didn't do properly when looking at the callers of this function. In fact, the [Code Review Comments wiki on empty slices](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices) explicitly states that it's idiomatic to prefer a `nil` slice over an empty slice in most cases. So apologies for the confusion here.

Instead, the callsite in `findAllDependencies` that tests `refs != nil` should instead be testing `len(refs) > 0`. That would obviate the need for the `nil` return. And this function body could look like:

```go
	var result []string
	matches := pattern.FindAllStringSubmatch(c.data, -1)
	for i, match := range matches {
		result = append(result, match[1])
	}
	return result
```

The function _may_ still return `nil` in the case that `len(matches) == 0`, but that's fine. Making the distinction between the nil slice and the empty slice in the `findAllDependencies` test results in additional logic here that ultimately isn't needed.

To answer the question on structs, this comment really only applies to pointer types, which include pointers, slices, maps, channels, functions, and interfaces. In Go, any type is implicitly initialized to its zero value. For pointer types, this is `nil`. Dave Cheney has a [good blog post on the zero value and making it useful](https://dave.cheney.net/2013/01/19/what-is-the-zero-value-and-why-is-it-useful).

The reason I look out for this sort of thing is that semantics of the zero value differ for different types, and dereferencing `nil` is a runtime error -- so that's potentially problematic for pointer types.

So why not for slices? Slices are often manipulated and accessed through builtin APIs. For example, ranging over a `nil` slice is the same as ranging over a 0-length slice: the range returns nothing. The `len` and `cap` of a `nil` slice will return 0. `Append`ing to a nil slice will magically create storage. This is why `nil` is fine in most cases for slices.

Maps, functions, and interfaces are a bit different. Maps don't have builtin APIs for most accesses; inserting into or reading from a `nil` map is a runtime error. Function-typed objects with `nil` values cannot be called; this is a runtime error. `Nil` interfaces are especially confusing; there's even a [FAQ on why a nil error value is not equal to nil](https://golang.org/doc/faq#nil_error) (though the text there applies to interfaces in general, it's often encountered when dealing with `error`, which is probably the most commonly used interface type).

This subtlety also extends to pointer types. Methods with pointer receivers (e.g. `func (t *T) Foo() error`) may expect that pointer to be non-`nil`. However, you may still call a method on a `nil`-valued type (e.g. `var t *T; t.Foo()`). If T is (say) a struct, and `(*T).Foo` accesses fields in that struct, this is also a runtime error.

So that's where this commentary is coming from, and things to watch out for when dealing with pointer types in Go.

Reply:
thanks for such deep explanation!